### PR TITLE
Dump tool for verifing get_json_object

### PR DESCRIPTION
**Pull request by res-life**
Verify the CPU and GPU results of get_json_object, if find diffs, save to Csv files.
This is only for testing.

- port Spark get_json_object code
- compare cpu/gpu results
- write to Csv file if have diffs. Append to csv file, one day create one csv file. By default, csv file name format is:  /tmp/get-json-object_diffs_yyyyMMdd.csv

How to open, set the following configs:
```
{
'spark.rapids.sql.expression.GetJsonObject': 'true',  // enable GetJsonObject
'spark.rapids.sql.test.get_json_object': 'true',  // enable verify GetJsonObject
'spark.rapids.sql.test.get_json_object.savePath': '/tmp/get-json-object_diffs_', // path will be /tmp/get-json-object_diffs_yyyyMMdd.csv, yyyyMMdd is current date. 
'spark.rapids.sql.test.get_json_object.saveRows': '1024' // max rows for each block in GetJsonObject operator when saving diffs. Note: this config can not control total rows, need tuning when the csv file is too big.
}
```

Signed-off-by: Chong Gao <res_life@163.com>

### CODE REVIEW

**Comment by res-life** [comment]
build

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/test/cpuJsonExpressions.scala [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+import com.nvidia.spark.rapids.{DumpUtils, GpuColumnVector, GpuScalar}
+import com.nvidia.spark.rapids.Arm.withResource
+
+import org.apache.spark.sql.catalyst.json.CreateJacksonParser
+import org.apache.spark.unsafe.types.UTF8String
+import org.apache.spark.util.Utils
+
+object CpuGetJsonObject {
+
+  private val maxRowsToSave = 1024 * 10
```
Comment (winningsix) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/test/cpuJsonExpressions.scala:37]: Let's make this configurable as an internal configuration.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala [hunk 2]
```diff
@@ -2114,6 +2114,20 @@ val SHUFFLE_COMPRESSION_LZ4_CHUNK_SIZE = conf("spark.rapids.shuffle.compression.
...
     .createOptional
 
+  val TEST_GET_JSON_OBJECT = conf("spark.rapids.sql.test.get_json_object")
+      .doc("Only for tests: verify for get_json_object. If enabled, dump the data that " +
+          "causes differences between Spark and GPU.")
+      .internal()
+      .booleanConf
+      .createWithDefault(false)
+
+  val TEST_GET_JSON_OBJECT_SAVE_PREFIX = conf("spark.rapids.sql.test.get_json_object.save.prefix")
```
Comment (winningsix) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala:2124]: nit: ```spark.rapids.sql.test.get_json_object.dump.file.dir``` better name?

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/test/cpuJsonExpressions.scala [hunk 3]
```diff
@@ -0,0 +1,419 @@
...
+                    }
+                  }
+                }
+              }
+            }
+          }
+
+          // 3. save table into a Parquet file
+          withResource(table) { _ =>
+            DumpUtils.dumpToParquetFile(table, verifyDiffSavePathPrefix)
```
Comment (winningsix) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/test/cpuJsonExpressions.scala:98]: can we just dump as a csv file?

**Review by winningsix — COMMENTED** [review]
Later, we need a doc introducing how this can be used.

**Comment by GaryShen2008** [comment]
Please target this to 24.04.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/test/cpuJsonExpressions.scala (hunk 1 above)
Comment (res-life) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/test/cpuJsonExpressions.scala]: Done.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala (hunk 2 above)
Comment (res-life) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala]: Now one day generates a file.
By default, csv file name format is: /tmp/get-json-object_diffs_yyyyMMdd.csv

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/test/cpuJsonExpressions.scala (hunk 3 above)
Comment (res-life) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/test/cpuJsonExpressions.scala]: Done.

**Comment by res-life** [comment]
build

**Comment by revans2** [comment]
I moved this to draft because I don't want to have this merged into an official. If we want to have something like this in an official release please talk to me directly and we can figure out what is the right way to add something like this.

**Review by revans2 — COMMENTED** [review]
I am seeing some problems with this where the CSV encoding does not handle new lines in the JSON data properly, and it also is getting confused on some inputs where the " escaping is causing all of the CSV parsers I tried to think there are more columns than there actually are. When I look at the output I can see that. Perhaps we need some more tests and it might be good to switch from CSV to TSV or some other separator character that is not likely to show up in JSON.

**Comment by res-life** [comment]
Haoyang is trying to read the CSV file via `com.univocity.parsers.csv.CsvReader`.
The CSV file which is wroten via `com.univocity.parsers.csv.CsvWriter` should be read properly by `com.univocity.parsers.csv.CsvReader`

**Pull request closed without merging**